README.md
# Pixelated Avatar Generator [![Travis CI Status](https://api.travis-ci.org/ExcaliburZero/pixelated-avatar-generator.svg)](https://travis-ci.org/ExcaliburZero/pixelated-avatar-generator) [![Coverage Status](https://coveralls.io/repos/github/ExcaliburZero/pixelated-avatar-generator/badge.svg?branch=master)](https://coveralls.io/github/ExcaliburZero/pixelated-avatar-generator?branch=master) [![Hackage](https://img.shields.io/hackage/v/pixelated-avatar-generator.svg)](https://hackage.haskell.org/package/pixelated-avatar-generator) [![pixelated-avatar-generator on Stackage Nightly](https://www.stackage.org/package/pixelated-avatar-generator/badge/nightly)](http://stackage.org/nightly/package/pixelated-avatar-generator)
Pixelated Avatar Generator is a Haskell library and application for generating pixelated avatar images from seed values.

```haskell
import Graphics.Avatars.Pixelated

createAndSaveAvatar :: String -> FilePath -> IO ()
createAndSaveAvatar s path = saveAvatar avatar path
  where avatar = scaleAvatar 32 $ generateAvatar seed
        seed   = createSeed s
```

![Some examples of avatars generated by the library.](example_image.png)

## Library
Pixelated Avatar Generator is a library which provides functions and data types for generating, altering, and saving pixelated avatars.

For information on how to use the library, see the documentation for the [`Graphics.Avatars.Pixelated`](https://hackage.haskell.org/package/pixelated-avatar-generator/docs/Graphics-Avatars-Pixelated.html) module.

### Documentation
The documentation for the latest release of the library can be found in [the library's Hackage entry](https://hackage.haskell.org/package/pixelated-avatar-generator).

Documentation of the development versions of the library can be generated by running the following command in the main directory of the project:

```
$ stack haddock
```

The generated html documentation can then be found in the following directory:

```
.stack-work/dist/x86_64-linux/Cabal-1.22.5.0/doc
```

## Executable
An executable program for generating avatar image files is also provided. It can generate several avatar images concurrently to given file locations.

The executable can be installed along with the library by running the following command:

```
$ stack install pixelated-avatar-generator
```

The executable can then by run by calling it with the desired filepath(s) of the output file(s) including the `.png` extension.

```
$ pixelated-avatar-generator image1.png image2.png
Successfully created 2 avatars.
```

By default, the avatars are at a size of 256x256px. Though the size can be changed by using a custom scaling factor via the `--scaling-factor` flag.

### Usage
```
Usage: pixelated-avatar-generator FILEPATH_1 [FILEPATH_2] [FILEPATH_3] ...

FILEPATH_(1...)   -- The locations to save generated avatars at. "img/test.png"

Options:

     --scaling-factor
         Use a custom scaling factor. The scaling factor is multiplied by 8 to
         get the dimensions of the image. For example, a scaling factor of 4
         would yield a 32x32px image. The default scaling factor is 32.
```

## License
The source code of Pixelated Avatar Generator is available under the [MIT license](https://opensource.org/licenses/MIT), see `LICENSE` for more information.


CHANGELOG.md
# Changelog

## 0.1.3
* Create new executable program.
* Make several small internal fixes.
* Add badges for Hackage and Stackage to readme.

## 0.1.2
* Add missing test files to Cabal config, to allow tests to pass from Hackage package.

## 0.1.1
* Implement saving with different image formats.
* Add image format conversion functions for PNG, GIF, and TIFF.
* Change the example image in the readme file.
* Add a link to the Hackage package to the readme file.
* Document the process of making a release of the package.
* Add information on the library to the readme file.
* Improve the documentation of the library.
* Improve some of the avatar colors.

## 0.1.0
* Implement generation and saving of avatars.
* Implement avatar upscaling.
* Create example executable program that makes use of the library.


test/Graphics/Avatars/PixelatedSpec.hs
module Graphics.Avatars.PixelatedSpec (main, spec) where

import Test.Hspec
import Test.QuickCheck

import Codec.Picture
import Data.String (IsString(..))
import qualified Data.ByteString.Lazy as B (ByteString(..), readFile)
import System.IO (hGetContents)
import System.IO.Unsafe (unsafePerformIO)

import Graphics.Avatars.Pixelated

-- `main` is here so that this module can be run from GHCi on its own.  It is
-- not needed for automatic spec discovery.
main :: IO ()
main = hspec spec

spec :: Spec
spec = do

  -----------------------------------------------------------------------------
  -- Seeds

  describe "createSeed" $ do
    it "creates a seed using an md5 checksum" $ do
      createSeed "" `shouldBe` Seed {unSeed = "d41d8cd98f00b204e9800998ecf8427e"}

  -----------------------------------------------------------------------------
  -- Avatars

  describe "Avatar" $ do
    it "can be compared for equality" $ do
      helloAvatar == helloAvatar `shouldBe` True
      helloAvatar /= testAvatar  `shouldBe` True
    it "can be converted into a string representation" $ do
      show helloAvatar `shouldBe` "Grey\n" ++ helloAvatarGridString

  describe "generateAvatar" $ do
    it "generates an avatar from a seed" $ do
      generateAvatar helloSeed `shouldBe` helloAvatar

  describe "scaleAvatar" $ do
    it "scales an avatar by a given factor" $ do
      scaleAvatar 2 testAvatar `shouldBe` testAvatar2x

  describe "convertAvatarToImage" $ do
    it "converts an avatar into an image" $ do
      encodePng (convertAvatarToImage helloAvatar) `shouldBe` helloAvatarImagePng

  ----------------------------------------
  -- Image Conversion

  describe "encodeToPng" $ do
    it "can encode an image into png format" $ do
      (encodeToPng . convertAvatarToImage) helloAvatar `shouldBe` helloAvatarImagePng

  describe "encodeToGif" $ do
    it "can encode an image into gif format" $ do
      (encodeToGif . convertAvatarToImage) helloAvatar `shouldBe` helloAvatarImageGif

  describe "encodeToTiff" $ do
    it "can encode an image into tiff format" $ do
      (encodeToTiff . convertAvatarToImage) helloAvatar `shouldBe` helloAvatarImageTiff

  -----------------------------------------------------------------------------
  -- Colors

  describe "colorFromSeed" $ do
    it "picks a color based on the given seed" $ do
      colorFromSeed (helloSeed) `shouldBe` Grey
    it "can choose Black" $ do
      colorFromSeed (Seed {unSeed = "c4000000000000000000000000000000"}) `shouldBe` Black
    it "can choose Blue" $ do
      colorFromSeed (Seed {unSeed = "ec000000000000000000000000000000"}) `shouldBe` Blue
    it "can choose Green" $ do
      colorFromSeed (Seed {unSeed = "a8000000000000000000000000000000"}) `shouldBe` Green
    it "can choose Grey" $ do
      colorFromSeed (Seed {unSeed = "aa000000000000000000000000000000"}) `shouldBe` Grey
    it "can choose Orange" $ do
      colorFromSeed (Seed {unSeed = "c8000000000000000000000000000000"}) `shouldBe` Orange
    it "can choose Purple" $ do
      colorFromSeed (Seed {unSeed = "c9000000000000000000000000000000"}) `shouldBe` Purple
    it "can choose Red" $ do
      colorFromSeed (Seed {unSeed = "c2000000000000000000000000000000"}) `shouldBe` Red
    it "can choose Yellow" $ do
      colorFromSeed (Seed {unSeed = "8f000000000000000000000000000000"}) `shouldBe` Yellow

  describe "getColorValue" $ do
    it "returns a pixel representation of a color" $ do
      getColorValue Orange `shouldBe` PixelRGB8 255 140 65
    it "can process Black" $ do
      getColorValue Black `shouldBe` PixelRGB8 41 41 41
    it "can process Blue" $ do
      getColorValue Blue `shouldBe` PixelRGB8 104 182 255
    it "can process Green" $ do
      getColorValue Green `shouldBe` PixelRGB8 114 220 131
    it "can process Grey" $ do
      getColorValue Grey `shouldBe` PixelRGB8 150 150 150
    it "can process Orange" $ do
      getColorValue Orange `shouldBe` PixelRGB8 255 140 65
    it "can process Purple" $ do
      getColorValue Purple `shouldBe` PixelRGB8 208 148 255
    it "can process Red" $ do
      getColorValue Red `shouldBe` PixelRGB8 255 87 87
    it "can process Yellow" $ do
      getColorValue Yellow `shouldBe` PixelRGB8 255 231 148

  -----------------------------------------------------------------------------
  -- Avatar Grids

  describe "generateAvatarGrid" $ do
    it "creates an avatar grid from a seed" $ do
      generateAvatarGrid helloSeed `shouldBe` helloAvatarGrid

  describe "AvatarGrid" $ do
    it "can be represented as a String" $ do
      show helloAvatarGrid `shouldBe` helloAvatarGridString

  -----------------------------------------------------------------------------
  -- Utilities

  describe "scaleList" $ do
   it "scales a list by a given factor" $ do
     scaleList 3 [0, 1] `shouldBe` [0, 0, 0, 1, 1, 1]
   it "can scale a list by a factor of one" $ do
     scaleList 1 [0, 1] `shouldBe` [0, 1]

-------------------------------------------------------------------------------
-- Values

helloSeed :: Seed
helloSeed = Seed {unSeed = "8b1a9953c4611296a827abf8c47804d7"}

helloAvatar :: Avatar
helloAvatar = Avatar {
      color = Grey
    , grid  = helloAvatarGrid
  }

helloAvatarGrid :: AvatarGrid
helloAvatarGrid = AvatarGrid ([
      [True, True, False, True, True, False, True, True]
    , [True, True, False, False, False, False, True, True]
    , [True, False, False, False, False, False, False, True]
    , [False, False, True, False, False, True, False, False]
    , [True, True, False, False, False, False, True, True]
    , [True, True, True, True, True, True, True, True]
    , [True, False, False, True, True, False, False, True]
    , [False, False, True, False, False, True, False, False]
  ])

helloAvatarGridString :: String
helloAvatarGridString = (init . unlines) [
    "██ ██ ██"
  , "██    ██"
  , "█      █"
  , "  █  █  "
  , "██    ██"
  , "████████"
  , "█  ██  █"
  , "  █  █  "
  ]

helloAvatarImagePng :: B.ByteString
helloAvatarImagePng = unsafePerformIO $ B.readFile "test/Graphics/Avatars/helloAvatar.png"

helloAvatarImageGif :: B.ByteString
helloAvatarImageGif = unsafePerformIO $ B.readFile "test/Graphics/Avatars/helloAvatar.gif"

helloAvatarImageTiff :: B.ByteString
helloAvatarImageTiff = unsafePerformIO $ B.readFile "test/Graphics/Avatars/helloAvatar.tiff"

testAvatar :: Avatar
testAvatar = Avatar {
      color = Orange
    , grid  = AvatarGrid [
         [True, False]
       , [False, True]
      ]
  }

testAvatar2x :: Avatar
testAvatar2x = Avatar {
      color = Orange
    , grid  = AvatarGrid [
         [True, True, False, False]
       , [True, True, False, False]
       , [False, False, True, True]
       , [False, False, True, True]
      ]
  }

-------------------------------------------------------------------------------
-- Utility Functions

unRight :: Either a b -> b
unRight x = case x of
  Right y -> y
  Left _  -> error "Unexpected Left"


app/Main.hs
module Main where

import Console.Options
import Control.Concurrent.Async (mapConcurrently)
import Data.Char (isDigit)
import Data.List (nub)
import Data.Maybe (fromMaybe)
import Data.Monoid ((<>))
import Data.Version (makeVersion)
import Graphics.Avatars.Pixelated
import System.Random (randomIO)

main :: IO ()
main = defaultMain $ do
  -- Define program information
  programName "pixelated-avatar-generator"
  programDescription "Generates pixelated avatar images"
  programVersion $ makeVersion [0,1,2]

  -- Gather program flags and arguments
  flagS <- flagParam
    (FlagLong "scaling-factor" <> FlagDescription scalingFlagDescription)
    (FlagRequired scaleParser)
  firstFile <- argument "FILEPATH" Right
  tailFiles <- remainingArguments "FILEPATHS"

  -- Perform the functionality of the program
  action $ \toParam -> do
    let scalingFactor = fromMaybe defaultScalingFactor (toParam flagS)
    let paths = nub $ toParam firstFile : toParam tailFiles
    _ <- mapConcurrently (saveRandomAvatar scalingFactor) paths
    putStrLn $ createSuccessMessage paths

-- | The default avatar scaling factor that is used when the user does not
-- specify a custom scaling factor via the "--scaling-factor" flag.
defaultScalingFactor :: Int
defaultScalingFactor = 32

-- | The description of the "--scaling-factor" flag.
scalingFlagDescription :: String
scalingFlagDescription = unwords
  [ "Use a custom scaling factor. The scaling factor is multiplied by 8 to get"
  , "the dimensions of the image. For example, a scaling factor of 4 would"
  , "yield a 32x32px image. The default scaling factor is 32."
  ]

-- | Prints a success message noting the number of images created.
--
-- >>> createSuccessMessage ["1.png", "2.png"]
-- "Successfully created 2 avatars."
createSuccessMessage :: [FilePath] -> String
createSuccessMessage paths = message
  where
    message = unwords ["Successfully created", numberOfAvatars, "avatars."]
    numberOfAvatars = (show . length) paths

-- | Creates and saves a random avatar at the given filepath using the given
-- scaling factor.
saveRandomAvatar :: Int -> FilePath -> IO ()
saveRandomAvatar factor path = do
  seed <- show <$> (randomIO :: IO Double)
  saveCustomAvatar factor path seed

-- | Creates and saves an avatar at the given filepath using the given seed
-- string and the given scaling factor.
saveCustomAvatar :: Int -> FilePath -> String -> IO ()
saveCustomAvatar factor path string = do
  let seed = createSeed string
  let avatar = generateAvatar seed
  let avatarScaled = scaleAvatar factor avatar
  saveAvatar avatarScaled path

-- | Attempts to parse a scaling factor, which must be a non-positive integer
-- (1, 2, 3, ...).
--
-- >>> scaleParser "1"
-- Right 1
-- >>> scaleParser "0"
-- Left "The given String is not a positive integer"
-- >>> scaleParser "-1"
-- Left "The given String is not a positive integer"
scaleParser :: String -> Either String Int
scaleParser string
    | isPosInt string = Right ((read :: String -> Int) string)
    | otherwise       = Left "The given String is not a positive integer"
  where
    isPosInt s   = isNaturalNum s && isNotZero s
    isNaturalNum = all isDigit
    isNotZero s  = not $ all (== '0') s


src/Graphics/Avatars/Pixelated.hs
-- |
-- Module      : Graphics.Avatars.Pixelated
-- Description : Contains types and functions for generating pixelated avatars.
-- Copyright   : (c) Christopher Wells, 2016
-- License     : MIT
-- Maintainer  : cwellsny@nycap.rr.com
--
-- This module provides types and functions for creating and working with
-- pixelated avatars.
--
-- Avatars can be generated by providing a `Seed`. Seeds can be created by
-- passing a random String into `createSeed`. The given String is then turned
-- into an MD5 checksum, which is used as the seed value.
--
-- Once a Seed has been created, an Avatar can be generated from it by passing
-- it into `generateAvatar`. By default, Avatars start at a size of 8x8px. The
-- size of an Avatar can be increased by passing it into `scaleAvatar`. Once
-- you have scaled the Avatar to the size you want it to be it can then be
-- saved to a file by passing it into `saveAvatar`.
--
-- By default, the `saveAvatar` function saves the avatar as a png image file,
-- however you can also save avatars in other image formats by using
-- `saveAvatarWith`. It allows you to specify the function to use to convert
-- the avatar image into an image ByteString.
--
-- = Example
-- The following is an example showing how to construct a function which will
-- generate a 256x256px avatar from a given seed string, and save it at the
-- given location.
--
-- @
-- import Graphics.Avatars.Pixelated
--
-- createAndSaveAvatar :: String -> FilePath -> IO ()
-- createAndSaveAvatar s path = saveAvatar avatar path
--   where avatar = scaleAvatar 32 $ generateAvatar seed
--         seed   = createSeed s
-- @
module Graphics.Avatars.Pixelated
(
  -- * Types
  -- ** Seed
  Seed(..), createSeed,

  -- ** Avatar
  Avatar(..), generateAvatar, scaleAvatar, saveAvatar, saveAvatarWith, convertAvatarToImage,

  -- *** Image Conversion
  ImageConversion, encodeToPng, encodeToGif, encodeToTiff,

  -- ** Color
  Color(..), getColorValue, colorFromSeed,
  
  -- ** Avatar Grid
  AvatarGrid(..), showGrid, generateAvatarGrid,

  -- ** Utility
  scaleList
)
where

import Codec.Picture (encodeColorReducedGifImage, encodePng, encodeTiff, generateImage, Image(..), PixelRGB8(..))
import Data.Char (ord)
import qualified Data.ByteString.Lazy as B (ByteString, writeFile)
import Data.Digest.Pure.MD5 (md5)
import Data.List.Split (chunksOf)
import Data.String (fromString)

-------------------------------------------------------------------------------
-- Seeds

-- | A seed to use in generating an avatar. Can be created from a String by
-- using the `createSeed` function.
--
-- Seeds are expected to be 32 character hexidecimal MD5 checksums.
newtype Seed = Seed { unSeed :: String }
  deriving (Eq, Show)

-- | Creates a seed from a given String.
--
-- >>> createSeed "Hello"
-- Seed {unSeed = "8b1a9953c4611296a827abf8c47804d7"}
createSeed :: String -> Seed
createSeed = Seed . show . md5 . fromString

-------------------------------------------------------------------------------
-- Avatars

-- | A generated avatar, comprised of a color and a grid representing the
-- visual pattern of the avatar image. Can be created from a seed using
-- `generateAvatar`.
--
-- Avatars are generated at a size of 8x8px, and can be scaled up to larger
-- image sizes by using the `scaleAvatar` function.
data Avatar = Avatar {
      color :: Color
    , grid  :: AvatarGrid
  }
  deriving (Eq)

-- | Generates a String containing the color and pattern of the avatar.
instance Show Avatar where
  show a = (show . color) a ++ "\n" ++ (show . grid) a

-- | Generates an avatar from the given seed.
--
-- >>> generateAvatar Seed {unSeed = "8b1a9953c4611296a827abf8c47804d7"}
-- Grey
-- ██ ██ ██
-- ██    ██
-- █      █
--   █  █  
-- ██    ██
-- ████████
-- █  ██  █
--   █  █  
generateAvatar :: Seed -> Avatar
generateAvatar seed = avatar
 where avatar = Avatar {
             color = aColor
           , grid = aGrid
         }
       aColor = colorFromSeed seed
       aGrid = generateAvatarGrid seed

-- | Scales the given Avatar by the given scaling factor.
--
-- For example, scaling an 8x8px avatar by a factor of 4 would transform it
-- into a 32x32px avatar.
scaleAvatar :: Int -> Avatar -> Avatar
scaleAvatar factor avatar = avatar { grid = AvatarGrid scaledGrid }
  where scaledGrid = (scaleList factor . map (scaleList factor)) unscaledGrid
        unscaledGrid = unAvatarGrid $ grid avatar

-- | Saves the given avatar as a png image file to the given file path. The
-- filepath should be the path and name of the image file to be created,
-- including the file extension.
--
-- For saving an avatar in a non-png format, use `saveAvatarWith`.
--
-- @
-- makeAvatar :: Seed -> FilePath -> IO ()
-- makeAvatar seed path = do
--   let avatar = generateAvatar seed path
--   saveAvatar avatar path
-- @
saveAvatar :: Avatar -> FilePath -> IO ()
saveAvatar = saveAvatarWith encodeToPng

-- | Saves the given avatar to the given file location, using the given
-- function to encode it into a specific image format.
--
-- Some examples of encoding functions are `encodeToGif` and `encodeToTiff`.
--
-- @
-- saveTiffAvatar :: Seed -> FilePath -> IO ()
-- saveTiffAvatar seed path = do
--   let avatar = generateAvatar seed path
--   saveAvatarWith encodeTiff avatar path
-- @
saveAvatarWith :: ImageConversion -> Avatar -> FilePath -> IO ()
saveAvatarWith conversion avatar path = do
  let image = conversion $ convertAvatarToImage avatar
  B.writeFile path image

-- | Converts the given Avatar into an Image.
convertAvatarToImage :: Avatar -> Image PixelRGB8
convertAvatarToImage avatar = image
  where image = generateImage getPixel dimension dimension
        dimension = length colorGrid
        getPixel x y = colorGrid !! y !! x
        colorGrid = (map . map) (toPixel $ color avatar) $ unAvatarGrid $ grid avatar
        toPixel c v = if v then getColorValue c else PixelRGB8 255 255 255

----------------------------------------
-- Image Conversion

-- | A function which converts an image into a lazy ByteString.
type ImageConversion = (Image PixelRGB8 -> B.ByteString)

-- | Converts an image into a Png image ByteString.
encodeToPng :: ImageConversion
encodeToPng = encodePng

-- | Converts an image into a Gif image ByteString.
encodeToGif :: ImageConversion
encodeToGif img = case encodeColorReducedGifImage img of
  Right i -> i
  Left  _ -> error "Unable to create valid gif color palette for avatar image."

-- | Converts an image into a Tiff image ByteString.
encodeToTiff :: ImageConversion
encodeToTiff = encodeTiff

-------------------------------------------------------------------------------
-- Colors

-- | A color of an avatar. The color of an avatar is the color that is applied
-- to the pattern of the avatar when it is converted into an image.
data Color = Black | Blue | Green | Grey | Orange | Purple | Red | Yellow
  deriving (Eq, Show, Enum)

-- | Converts the given color into a RGB pixel representation.
--
-- >>> getColorValue Orange
-- PixelRGB8 255 140 65
getColorValue :: Color -> PixelRGB8
getColorValue c
  | c == Black  = PixelRGB8 41  41  41
  | c == Blue   = PixelRGB8 104 182 255
  | c == Green  = PixelRGB8 114 220 131
  | c == Grey   = PixelRGB8 150 150 150
  | c == Orange = PixelRGB8 255 140 65
  | c == Purple = PixelRGB8 208 148 255
  | c == Red    = PixelRGB8 255 87  87
  | otherwise   = PixelRGB8 255 231 148

-- | Picks an avatar color using the given seed.
--
-- Each of the eight possible colors has roughly an equal chance of being
-- chosen with a random seed.
--
-- >>> colorFromSeed $ Seed {unSeed = "8b1a9953c4611296a827abf8c47804d7"}
-- Grey
colorFromSeed :: Seed -> Color
colorFromSeed = genColor . dSum . unSeed
  where twoDigits n = map ord $ take 2 n
        dSum n = foldr (+) 1 $ twoDigits n
        genColor a = [Black .. Yellow] !! (a `mod` 8)

-------------------------------------------------------------------------------
-- AvatarGrids

-- | A grid of boolean values representing an Avatar. True values indicate
-- colored pixels, and False values indicate blank pixels.
newtype AvatarGrid = AvatarGrid { unAvatarGrid :: [[Bool]] }
  deriving (Eq)

-- | Converts the grid into a String representation.
instance Show AvatarGrid where
  show = showGrid . unAvatarGrid

-- | The left half of an AvatarGrid.
newtype AvatarGridSide = AvatarGridSide { unAvatarGridSide :: [[Bool]] }

-- | Converts the grid side into a String representation.
instance Show AvatarGridSide where
  show = showGrid . unAvatarGridSide

-- | Converts a grid of boolean values into a String representation.
--
-- >>> putStrLn $ showGrid [[True, False], [False, True]]
-- █ 
--  █
showGrid :: [[Bool]] -> String
showGrid g = (init . unlines) $ (map . map) showPixel g

-- | Converts a boolean value into a character representation for a pixel.
--
-- >>> showPixel True
-- '\9608'
-- >>> showPixel False
-- ' '
showPixel :: Bool -> Char
showPixel p = if p then '█' else ' '

-- | Generates an AvatarGrid using the given Seed.
--
-- It works by generating the left half of the grid using the contents of the
-- Seed, and then mirroring the left half to create the full grid.
--
-- >>> generateAvatarGrid Seed {unSeed = "8b1a9953c4611296a827abf8c47804d7"}
-- ██ ██ ██
-- ██    ██
-- █      █
--   █  █  
-- ██    ██
-- ████████
-- █  ██  █
--   █  █  
generateAvatarGrid :: Seed -> AvatarGrid
generateAvatarGrid = mirrorGrid . generateAvatarGridSide

-- | Creates a full AvatarGrid by mirroring the given AvatarGridSide on the
-- y-axis.
mirrorGrid :: AvatarGridSide -> AvatarGrid
mirrorGrid side = AvatarGrid $ map mirror $ unAvatarGridSide side
  where mirror l = l ++ reverse l

-- | Generates the right side of an AvatarGrid using the given seed.
generateAvatarGridSide :: Seed -> AvatarGridSide
generateAvatarGridSide = AvatarGridSide . numToGrid . unSeed

-- | Converts the given hexidecimal number String into a grid of boolean values.
numToGrid :: String -> [[Bool]]
numToGrid s = boolGrid
  where boolGrid = (map . map) (convertToPixel . ord) numGrid
        numGrid  = chunksOf 4 s
        convertToPixel = (> ord '7')

-------------------------------------------------------------------------------
-- Utilities

-- | Scales the given list by the given scaling factor.
--
-- >>> scaleList 2 [1, 2]
-- [1,1,2,2]
-- >>> scaleList 3 [0, 1]
-- [0,0,0,1,1,1]
scaleList :: Int -> [a] -> [a]
scaleList _     []     = []
scaleList factor (x:xs) = replicate factor x ++ scaleList factor xs
